DOCSP-30560: Connection troubleshooting (#38)
mongodb/docs-rust: source/connection-troubleshooting.txt

   - The :ref:`Issues & Help <rust-issues-and-help>` page, which has
     information about reporting bugs, contributing to the driver, and 
     finding more resources
   - The `MongoDB Community Forums <https://community.mongodb.com>`__ for
     questions, discussions, or general technical support

   ..
       TODO:
     - The :ref:`Frequently Asked Questions (FAQ) <rust-faq>` for the
       {+driver-short+}
     - The :ref:`Operation Error Handling <rust-operation-errors>` page
       for the {+driver-short+}
      

Server Connection Error
~~~~~~~~~~~~~~~~~~~~~~~

When you attempt to connect to a server, the {+driver-short+} might generate an
error message. This message might look similar to the following message indicating
that the driver cannot connect to a server on the specified hostname or port:

.. code-block:: none
   :copyable: false

   Error: Error { kind: ServerSelection { message: "Server selection timeout: 
   No available servers. Topology: { Type: Unknown, Servers: [ { Address: 
   127.0.0.1:27017, Type: Unknown, Error: Kind: I/O error: Connection refused 
   (os error 61), labels: {} } ] }" }, labels: {}, wire_version: None, source: 
   None }

> The following sections describe methods that may help resolve the issue.

Reviewer: S: replace may with might ([SG](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#:~:text=than%20matrices.-,may,-Use%20may%20to))

the authorization is not configured correctly. In these cases, the driver
raises an error message similar to the following message:

.. code-block:: none
   :copyable: false

    Error: Error { kind: Authentication { message: "SCRAM failure: bad auth : 
    authentication failed" }, labels: {}, wire_version: None, source: Some(Error 
    { kind: Command(CommandError { code: 8000, code_name: "AtlasError", message:
    "bad auth : authentication failed", topology_version: None }),
    labels: {}, wire_version: None, source: None }) }

The following sections describe methods that may help resolve the issue.

.. _rust-troubleshooting-connection-string-auth:

Check Connection String
-----------------------

An invalid connection string is the most common cause of authentication
issues when attempting to connect to MongoDB.

.. tip::

   For more information about connection strings,
   see the :ref:`rust-quick-start-connection-string` guide.

If your connection string contains a username and password, ensure that they
are in the correct format.

.. note::

   If your username or password includes any of the following characters, you

>    must `percent encode <https://tools.ietf.org/html/rfc3986#section-2.1>`__ them:

Reviewer: In this note, it might also be helpful to tell users what to do with the percent encoded values. For example adding a sentence like "You can then use your percent encoded username and password as credentials in your connection string."

   127.0.0.1:27017, Type: Unknown, Error: Kind: I/O error: Connection refused 
   (os error 61), labels: {} } ] }" }, labels: {}, wire_version: None, source: 
   None }

The following sections describe methods that might help resolve the issue.

.. _rust-troubleshooting-connection-string-port:

Check Connection String
-----------------------

Verify that the hostname and port number in the connection string are both
accurate. In the sample error message, the hostname is ``127.0.0.1`` and the
port is ``27017``. The default port value for a MongoDB instance is
``27017``, but you can configure MongoDB to communicate on another port.

.. _rust-troubleshooting-connection-firewall:

Configure Firewall
------------------

Assuming that your MongoDB deployment uses the default port, verify that port
``27017`` is open in your firewall. If your deployment uses a different port,
verify the correct port is open in your firewall.

.. warning::

   Do not open a port in your firewall unless you are sure that it is the port
   used by your MongoDB instance.

.. _rust-troubleshooting-connection-number-connections:

Check the Number of Connections
-------------------------------

> Each ``MongoClient`` instance supports a maximum number of concurrent open

Reviewer 2: Suggest using `Client` rather than `MongoClient` to match Rust driver API [Client](https://docs.rs/mongodb/latest/mongodb/struct.Client.html).

.. rust-connection-troubleshooting:

==========================
Connection Troubleshooting
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

This page offers potential solutions to issues you might encounter when
using the {+driver-long+} to connect to a MongoDB deployment.

.. note::

   This page only addresses connection issues. If you encounter any other issues
   with MongoDB or the driver, visit the following resources:

   - The :ref:`Issues & Help <rust-issues-and-help>` page, which has
     information about reporting bugs, contributing to the driver, and 
     finding more resources
   - The `MongoDB Community Forums <https://community.mongodb.com>`__ for
     questions, discussions, or general technical support

   ..
     - The :ref:`Frequently Asked Questions (FAQ) <rust-faq>` for the

>        {+driver-short+}

Reviewer: S: add another TODO here to link to the operation error page thats in progress

Authentication Error
~~~~~~~~~~~~~~~~~~~~

The {+driver-short+} may be unable connect to a MongoDB instance if
the authorization is not configured correctly. In these cases, the driver
raises an error message similar to the following message:

.. code-block:: none
   :copyable: false

    Error: Error { kind: Authentication { message: "SCRAM failure: bad auth : 
    authentication failed" }, labels: {}, wire_version: None, source: Some(Error 
    { kind: Command(CommandError { code: 8000, code_name: "AtlasError", message:
    "bad auth : authentication failed", topology_version: None }),
    labels: {}, wire_version: None, source: None }) }

The following sections describe methods that may help resolve the issue.

.. _rust-troubleshooting-connection-string-auth:

Check Connection String
-----------------------

An invalid connection string is the most common cause of authentication
issues when attempting to connect to MongoDB.

.. tip::

   For more information about connection strings,
   see the :ref:`rust-quick-start-connection-string` guide.

If your connection string contains a username and password, ensure that they
are in the correct format.

.. note::

>    If the username or password includes any of the following characters, they
>    must be `percent encoded <https://tools.ietf.org/html/rfc3986#section-2.1>`__:

Reviewer: ```suggestion
   If your username or password includes any of the following characters, you must
   `percent encoded <https://tools.ietf.org/html/rfc3986#section-2.1>`__ it:
```

   .. code-block:: none

      : / ? # [ ] @

When connecting to a replica set, include all the replica set hosts in
your connection string. Separate each of the hosts in the connection
string with a comma. This enables the driver to establish a connection if

> one of the hosts is unreachable.

Reviewer: S: could we include an example of what this looks like or link to an example? Theres an example in the connection guide: https://docs-mongodbcom-staging.corp.mongodb.com/rust/docsworker-xlarge/DOCSP-30560-connection-troubleshoot/fundamentals/connections/connection-guide/#connect-to-a-replica-set

.. _rust-troubleshooting-connection-auth-mechanism:

Verify the Authentication Mechanism
-----------------------------------

Ensure that your credentials and authentication mechanism are correct. You can
specify your authentication credentials in the options of your connection string
or in a ``Credential`` struct.

To learn more about authentication, see the
:ref:`rust-authentication` guide.

.. _rust-troubleshooting-connection-admin:

Verify User Is in Authentication Database
-----------------------------------------

To successfully authenticate a connection by using a username and password,
the username must be defined in the authentication database. The default
authentication database is the ``admin`` database. To use a different database

> for authentication, specify the ``authSource`` in the connection string. The

Reviewer: ```suggestion
for authentication, specify the ``authSource`` option in the connection string. The
```

following example instructs the driver to use ``users`` as the authentication
database:

.. code-block:: rust
   :copyable: true

   let uri = "mongodb://<username>:<password>@<hostname>:<port>/?authSource=users";
   let client = Client::with_uri_str(uri).await?;

Error Sending a Command
~~~~~~~~~~~~~~~~~~~~~~~

When the driver fails to send a command after you make a request,
it raises an error message similar to the following message:

.. code-block:: none
   :copyable: false

   Error { kind: Command(CommandError { code: 17312, code_name: "Location17312",

>    message: "sample message" }), labels: {} }', src\main.rs:00:00

Reviewer: I: I'm not sure that this error type fits the description given in the introduction. From the API docs, it looks like CommandError is raised when a database command fails, not when the driver fails to send a command. Could you follow up with eng about this?

Author: Yes I did some more research on this and you're right -- what do you think about just taking out this section? it basically just links to other sections on the page anyway
